sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -2882,6 +2882,31 @@ class DataFrameSuite extends QueryTest
     df2.collect()
     assert(accum.value == 15)
   }
+
+  test("SPARK-35560: Remove redundant subexpression evaluation in nested subexpressions") {
+    Seq(1, Int.MaxValue).foreach { splitThreshold =>
+      withSQLConf(SQLConf.CODEGEN_METHOD_SPLIT_THRESHOLD.key -> splitThreshold.toString) {
+        val accum = sparkContext.longAccumulator("call")
+        val simpleUDF = udf((s: String) => {
+          accum.add(1)
+          s
+        })
+
+        // Common exprs:
+        //  1. simpleUDF($"id")
+        //  2. functions.length(simpleUDF($"id"))
```

Comment:
Q: What if a tree has more deeply-nested common exprs? The current logic can work well? e.g., I thought it like this;
```
        // subExpr1 = simpleUDF($"id");
        // subExpr2 = functions.length(subExpr1);
        // subExpr3 = functions.xxxx(subExpr2);
        // subExpr4 = ...
```

Reply:
Yes, this is actually the cases this logic to deal with. Previous common expression gen-ed codes will put into the map. The code generator looks up into the map when generating code for later common expressions to replace the semantic-equal expression with gen-ed code value.

Reply:
Nice